Added Static CustomOp with lowering to regular custom Op

mehrdad2m opened the pull request:
**Context:**
Currently, the classical parameters of quantum gates are compiled dynamically even if they are known at compile time. We would like the IR to be extended to support literal values as opposed to SSA Values for such static parameters.

**Description of the Change:**
This is achieved by adding a new type of gate called StaticCustomOp which is similar to CustomOp except that it uses a DenseF64ArrayAttr which lists all the literal values for parameters in square bracket. For example the following IR: 
```
%c = llvm.mlir.constant(2.000000e-01 : f64)
%result = quantum.custom "RX"(%c) %qubit
```
would change to:
```
%result = quantum.static_custom "RX" [2.000000e-01] %qubit
```
The static custom ops are then lowered to the regular custom ops after apply-transform-sequence pass for the rest of the compilation process.
Currently we have **not** supported Multirz, and GlobalShift.

**Benefits:**
More flexibility in case there is a lack of support for dynamic quantum circuits.

**Possible Drawbacks:**

**Related GitHub Issues:**
Static parameters [sc-73581]

mehrdad2m commented:
> This looks a lot better! You'll also need to change the merge rotations and cancel inverses passes right?

Thanks @erick-xanadu, yes I am planning to change those today.

mehrdad2m commented on mlir/lib/Quantum/IR/QuantumOps.cpp:
```diff
@@ -40,6 +40,35 @@ static const mlir::StringSet<> hermitianOps = {"Hadamard", "PauliX", "PauliY", "
+
+LogicalResult StaticCustomOp::canonicalize(StaticCustomOp op, mlir::PatternRewriter &rewriter)
+{
+    if (op.getAdjoint()) {
```
Yes I agree with this. I think I just copied over customOp canonicalize method here. Didn't look too much into it :) I will fix this as well

dime10 replied:
> > This looks a lot better! You'll also need to change the merge rotations and cancel inverses passes right?
> Thanks @erick-xanadu, yes I am planning to change those today.

I also don't think these need to be updated to run on the static ops since they can run after the conversion to dynamic.

mehrdad2m replied:
With the current design, the lowering is done after the apply-transform-sequence pass which is ultimately the place where peephole optimizations are run, so passes like cancel inverses and merge rotations should be able to deal with static custom ops. I think the reason for this placement in pipeline is to allow external contributors to schedule their passes within the apply-transform-sequence pass.

dime10 replied:
Ah I see, oh well then I guess it's fine 👍

dime10 commented on frontend/catalyst/jax_extras/tracing.py:
```diff
@@ -985,6 +985,13 @@ def bind_flexible_primitive(primitive, flexible_args: dict[str, Any], *dyn_args,
     for flex_arg_name, flex_arg_value in flexible_args.items():
         if type(flex_arg_value) in static_literal_pool:
             static_args |= {flex_arg_name: flex_arg_value}
+        elif isinstance(flex_arg_value, list) and flex_arg_name == "static_params":
+            if not flex_arg_value:
+                continue
+            if all(type(arg) in static_literal_pool for arg in flex_arg_value):
+                static_args |= {flex_arg_name: flex_arg_value}
```
I wonder if we just allow lists in general here? Would avoid including this special case with the `static_params` name. 
Also some minor simplification of the code (no need for dictionary update syntax with only 1 element, can remove early exit since branch is still legal & inexpensive if list is empty):
```suggestion
            static_arg[flex_arg_name] = flex_arg_value
        elif isinstance(flex_arg_value, list):
            if all(type(arg) in static_literal_pool for arg in flex_arg_value):
                static_args[flex_arg_name] = flex_arg_value
```

dime10 replied:
Either let's update the docstring of this function to reflect the changes :)

mehrdad2m replied:
I applied the suggestions except for the early exit. if I don't have the early exit, I would get a {'static_param' : []} in the static arguments which causes some tests that compare JAXPRs to fail.

dime10 reviewed (changes requested):
Nice work! I know we also discussed the MultiRZ/GlobalPhase handling in the frontend so I'll wait for those for final approval, but otherwise nothing major 👍

mehrdad2m commented on frontend/catalyst/from_plxpr.py:
```diff
@@ -305,7 +305,7 @@ def interpret_operator_eqn(self, eqn: jax.core.JaxprEqn) -> None:
                 *invals,
                 op=eqn.primitive.name,
-                params_len=len(eqn.invars) - eqn.params["n_wires"],
+                ctrl_value_len=0,
```
the reason for this change is that after using using bind_flexible_primitive, the gate parameters will be appended to the list of all dynamic parameters and hence it will change the order of arguments. previously it was [qubits, params, ctrl_qubits, ctrl_values], but after the change it is [qubits, ctrl_qubits, ctrl_values, params]. so we need to know the length of ctrl_values instead of params.

dime10 replied:
Oh I see, the plxr converter doesn't support dynamic control qubits yet, so the 0 makes sense.

I just realized though that `ctrl_len` and `ctrl_value_len` should always be the same, so we might not need this variable and/or assert that it they are the same.

dime10 replied:
Still feel weird about the `ctrl_value_len` parameter being here, semantically it just feels out of place / not that meaningful. Can we just get rid of it?

mehrdad2m replied:
oh I just remembered that I added this to pass a test. There are some tests in `frontend/test/pytest/test_from_plxpr.py` that compare the jaxpr from plpxr and catalyst. If I remove the `ctrl_value_len` then that test fails because one will have the `ctrl_value_len` in the jaxpr and the other will not. But otherwise I agree that there is not point of having it there.

mehrdad2m commented on frontend/catalyst/pipelines.py:
```diff
@@ -169,6 +169,8 @@ def get_enforce_runtime_invariants_stage(_options: CompileOptions) -> List[str]:
         "split-multiple-tapes",
         # Run the transform sequence defined in the MLIR module
         "builtin.module(apply-transform-sequence)",
+        # Lower the static custom ops to regular custom ops with dynamic parameters.
+        "static-custom-lowering",
```
As far as I know, we expect changes from mqt to happen in apply-transform-sequence pass. That's why it is placed right after that.

dime10 replied:
Oh good point, that's where the plugins are applied. All good then 👍

mehrdad2m commented on mlir/include/Quantum/IR/QuantumOps.td:
```diff
@@ -456,6 +456,42 @@ def CustomOp : UnitaryGate_Op<"custom", [DifferentiableGate, NoMemoryEffect,
+    );
+
+    let assemblyFormat = [{
+        $gate_name $static_params $in_qubits attr-dict ( `ctrls` `(` $in_ctrl_qubits^ `)` )?  ( `ctrlvals` `(` $in_ctrl_values^ `)` )? `:` type($out_qubits) (`ctrls` type($out_ctrl_qubits)^ )?
```
Yes, the brackets would be printed anyways. I didn't spend time to make a custom printer, but will consider it if I have time before the end of week

dime10 replied:
Don't worry I think it's alright 👍

mehrdad2m commented on mlir/lib/Quantum/Transforms/StaticCustomPatterns.cpp:
```diff
@@ -0,0 +1,61 @@
+    {
+        LLVM_DEBUG(dbgs() << "Lowering the following static custom operation:\n" << op << "\n");
+        SmallVector<Value, 4> paramValues;
+        auto staticParams = op.getStaticParams();
+        for (auto param : staticParams) {
```
Oh Actually `op.getStaticParams() ` returns a `llvm::ArrayRef<double>` instead of the attribute which is convenient for looping over :) to get the attribute one should use op.getStaticParamsAttr();

dime10 replied:
Oh neat, thanks :)

mehrdad2m commented:
> Nice work! I know we also discussed the MultiRZ/GlobalPhase handling in the frontend so I'll wait for those for final approval, but otherwise nothing major 👍

Thanks @dime10 for the comments. I added two lit tests for the lowering pass and addressed all of the comments except for "MultiRZ/GlobalPhase handling in the frontend". I will work on it next, but maybe it is safer to do that in another PR? just to make sure this one get merged by the end of the week. what do you think?

dime10 commented on mlir/lib/Quantum/Transforms/static_custom_lowering.cpp:
```diff
@@ -0,0 +1,66 @@
+        ConversionTarget target(context);
+
+        target.addIllegalOp<StaticCustomOp>();
+        target.markUnknownOpDynamicallyLegal([](Operation *) { return true; });
```
Don't think this is necessary for a partial conversion, since it will only fail if explicitly illegal ops cannot be converted.
```suggestion
```

mehrdad2m replied:
target.markUnknownOpDynamicallyLegal was needed since the newly added `custom` and `arith.constant` operations that replace the `static_ustom` operation would fail to legalize. I removed this and legalized those two operations explicitly to avoid confusion.

dime10 replied:
Thanks, that's helpful!

dime10 reviewed (approved):
Nice work 💯 Happy for the odd gates to come in a follow up 👍
